Fixes for docker image and container build - Issue #31

BBBThunda opened the pull request:
* Change permissions on `/var/www` in the container so `composer` user can write to the `vendor/` directory.

* Add `zip` and `unzip` to the container dependencies because they are needed by `composer`

* Create `.env` from `.env.example` in the container since `.env` is in the `.dockerignore`

* Remove mount from `docker-compose` because the files are already copied into the image during build and mounting them wipes out all of the build changes.

BBBThunda committed:
Fix permission issues, don't mount code after build

BBBThunda commented on Dockerfile:
```diff
@@ -36,18 +44,19 @@ USER composer
+
 # Generate App Key
-RUN php artisan key:generate
+#RUN php artisan key:generate
```
Probably should uncomment this one... I'll test it out.

ramonacat commented on Dockerfile:
```diff
@@ -36,18 +44,19 @@ USER composer
 RUN /usr/local/bin/composer install
 
+# Since .env is in .dockerignore, we have to get it from somewhere
+RUN cp .env.example .env
```
no need to do this, docker-compose will read up the `.env` file, and the environment variables can be set based on that (i.e. same way they're set for the db here: https://github.com/erikaheidi/hacktober-board/blob/master/docker-compose.yml#L23-L24)

BBBThunda replied:
You are correct, with a `.env` file on your local machine, `docker-compose` will apply the environment variables during the build. (this should be in the readme, by the way. I'll add it.)

But without an `.env` file on the app container, post-build commands like `php artisan migrate` fail - I think because without the `.env` it defaults to `config/database.php` for the environment variables and you end up with this:

```
$ docker-compose exec app php artisan migrate
**************************************
*     Application In Production!     *
**************************************

 Do you really wish to run this command? (yes/no) [no]:
 > yes


   Illuminate\Database\QueryException  : SQLSTATE[HY000] [2002] Connection refused (SQL: select * from information_schema.tables where table_schema = forge and table_name = migrations and table_type = 'BASE TABLE')

  at /var/www/vendor/laravel/framework/src/Illuminate/Database/Connection.php:665
    661|         // If an exception occurs when attempting to run a query, we'll format the error
    662|         // message to include the bindings with SQL, which will make this exception a
    663|         // lot more helpful to the developer instead of just the database's errors.
    664|         catch (Exception $e) {
  > 665|             throw new QueryException(
    666|                 $query, $this->prepareBindings($bindings), $e
    667|             );
    668|         }
    669|

  Exception trace:

  1   PDOException::("SQLSTATE[HY000] [2002] Connection refused")
      /var/www/vendor/laravel/framework/src/Illuminate/Database/Connectors/Connector.php:70

  2   PDO::__construct("mysql:host=127.0.0.1;port=3306;dbname=forge", "forge", "", [])
      /var/www/vendor/laravel/framework/src/Illuminate/Database/Connectors/Connector.php:70

  Please use the argument -v to see more details.
```

 Maybe it's not the ideal solution, but when I added the `cp` command to the Dockerfile everything worked as expected.

BBBThunda replied:
Ahh, just realized it was the `php artisan key:generate` command that was failing initially and caused me to add the `cp` command. I wasn't seeing the error because I commented that line out. Now that I've uncommented it, I'm getting the exception again.

It looks like `php artisan key:generate` is explicitly requiring a `.env` file inside the container and fails if it doesn't find it, which breaks the docker build.

```
Step 16/18 : RUN php artisan key:generate
 ---> Running in f5e2c63c6c71

   ErrorException  : file_get_contents(/var/www/.env): failed to open stream: No such file or directory

  at /var/www/vendor/laravel/framework/src/Illuminate/Foundation/Console/KeyGenerateCommand.php:96
     92|     {
     93|         file_put_contents($this->laravel->environmentFilePath(), preg_replace(
     94|             $this->keyReplacementPattern(),
     95|             'APP_KEY='.$key,
  >  96|             file_get_contents($this->laravel->environmentFilePath())
     97|         ));
     98|     }
     99|
    100|     /**

  Exception trace:

  1   file_get_contents("/var/www/.env")
      /var/www/vendor/laravel/framework/src/Illuminate/Foundation/Console/KeyGenerateCommand.php:96

  2   Illuminate\Foundation\Console\KeyGenerateCommand::writeNewEnvironmentFileWith("base64:HN0NuNLB/+dL64ZPEjA4Y9c1aUSeYGGl9nGAg+7P2IY=")
      /var/www/vendor/laravel/framework/src/Illuminate/Foundation/Console/KeyGenerateCommand.php:80

  Please use the argument -v to see more details.
ERROR: Service 'app' failed to build: The command '/bin/sh -c php artisan key:generate' returned a non-zero code: 1
```

@Agares I'm open to suggestions if there's a better way to get around this, but simply removing it will un-fix the broken build issue that started me down this path in the first place.

My thinking is because this is a dev environment we could just remove `.env` from `.dockerignore`. Is it there for any particular reason?

erikaheidi replied:
They `key:generate` must happen after the `.env` is generated. I don't think the `.env` should be included in the image, do you?

BBBThunda replied:
Good point. I was kind of thinking that, but I figured since it's dev, maybe nobody cares.

If we want to remove the `.env` from the image, we could remove the `artisan` commands that rely on this file from the docker build and add them in an entrypoint script along with adding/generating the `.env` in the already running container.

Except we would still need a way to get around the fact that `.env` is in the `.dockerignore`. It doesn't get mounted, and I don't know if we can just copy it directly.

BBBThunda committed:
Add volume back with empty vendor volume

BBBThunda committed:
Uncomment key:generate and move USER step up

Removed commented out migrate and db:seed commands because those can't be run
until the db container is up, so they would probably make more sense in an
entrypoint script or something. They're referenced in readme.md anyway.

BBBThunda committed:
Add vendor/ to .dockerignore

Without this, docker build generates an empty vendor/ directory locally

BBBThunda commented:
Hi, any updates on this?

I refactored the Dockerfile in a separate branch to take advantage of docker image caching to speed up development and minimize the need for `composer install`s, etc., but I'd prefer to finish this up first so I can avoid conflicts, etc.

erikaheidi commented on Dockerfile:
```diff
@@ -36,18 +44,15 @@ USER composer
 RUN /usr/local/bin/composer install
 
-# Generate App Key
-RUN php artisan key:generate
```
After creating a `.env` file copied from the `.env.example` file, you still need to set up an App key. Maybe if you bring the `cp` command to this point, before generating the key, it will fix it?

BBBThunda replied:
See my response to your other comment: https://github.com/erikaheidi/hacktober-board/pull/32#discussion_r339569070

I will try this for now and if it works I'll update the branch and we can decide whether to leave it as-is for now or if we would rather pull the key generation into an entrypoint script.

BBBThunda replied:
I just realized I'm already running the key generation after the `cp` command. See here:
https://github.com/erikaheidi/hacktober-board/pull/32/files#diff-3254677a7917c6c01f55212f86c57fbfR54

Is that what you meant? Or am I misunderstanding something?

If this is correct, then I've tested it this way and everything seems to work fine.

erikaheidi replied:
yes that's exactly it, sorry I didn't notice it was there. thank you, I'll merge it now!

erikaheidi reviewed (changes requested):
Could you try this change to confirm it works? Ideally we need it to automatically generate this app key, it would make things easier.
